src/Lecture2.hs

```diff
@@ -7,8 +7,316 @@ Stability               : Stable
+  * Stomachs of green dragons contain extreme acid and they melt any
+    treasure except gold. So green dragons has only gold as reward.
+    All other dragons always contain treasure in addition to gold.
+  * Knight tries to slay dragon with their sword. Each sword strike
+    decreases dragon health by the "sword attack" amount. When the
+    dragon health becomes zero or less, a dragon dies and the knight
+    takes the reward.
+  * After each 10 sword strikes, dragon breathes fire and decreases
+    knight health by the amount of "dragon fire power". If the
+    knight's health becomes 0 or less, the knight dies.
+  * Additionally, each sword strike decreases "knight's endurance" by one.
+    If knight's endurance becomes zero, they become tired and are not
+    able to continue the fight so they run away.
+
+Implement data types to describe treasure, knight and dragon.
+And implement a function that takes a knight and a dragon and returns
+one of the three possible fight outcomes.
+
+You're free to define any helper functions.
+
+🕯 HINT: If you find the description overwhelming to implement entirely
+  from scratch, try modelling the problem in stages.
+
+    1. Implement all custom data types without using polymorphism.
+    2. Add @newtype@s for safety where you think is appropriate.
+    3. Encode the fight result as a sum type.
+    4. Add polymorphism.
+    5. Make all invalid states unrepresentable. Think, how you can
+       change your types to prevent green dragons from having any
+       treasure besides gold (if you already haven't done this).
+-}
+
+-- some help in the beginning ;)
+data Knight = Knight
+    { knightHealth    :: Int
+    , knightAttack    :: Int
+    , knightEndurance :: Int
+    }
+
+dragonFight = error "TODO"
+
+----------------------------------------------------------------------------
+-- Challenges
+----------------------------------------------------------------------------
+
+{- The following exercises are considered more challenging. However,
+you still may find some of them easier than some of the previous
+ones. Difficulty is a relative concept.
+-}
+
+{- | Write a function that takes a list of numbers and returns 'True'
+if all the numbers are in the increasing order (i.e. the list is
+sorted).
+
+>>> isIncreasing [3, 1, 2]
+False
+>>> isIncreasing [1 .. 10]
+True
+-}
+isIncreasing :: [Int] -> Bool
+isIncreasing [] = True
+isIncreasing [_] = True
+isIncreasing (x : y : xs) = x < y && isIncreasing (y:xs)
+{- | Implement a function that takes two lists, sorted in the
+increasing order, and merges them into new list, also sorted in the
+increasing order.
+
+The lists are guaranteed to be given sorted, so you don't need to
+verify that.
+
+>>> merge [1, 2, 4] [3, 7]
+[1,2,3,4,7]
+-}
+merge :: [Int] -> [Int] -> [Int]
+merge [] [] = []
+merge [] xs = xs
+merge xs [] = xs
+merge (x:xs) (y:ys) = if x < y
+                        then x : merge xs (y : ys)
+                        else y : merge (x : xs) ys
```

Comment:
Absolutely correct solution! 💎 
You can make it more performant by checking if both heads of the given lists are equal. In that case, you're able to reduce the size of each list by 1 on each recursive call instead of reducing the size of only a single list ✂️

Reply:
lile this ?
| x == y = merge xs ys

Reply:
Almost 🙂  Don't forget to add elements!

```haskell
| x = y = x : y : merge xs ys
```